File: 백준/Silver/2529. 부등호/부등호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    static int k;
    static String[] arr;
    static List<String> answer;
    static boolean[] visited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        k = Integer.parseInt(br.readLine()) + 1;
        arr = br.readLine().split(" ");
        answer = new ArrayList<>();
        visited = new boolean[10];
        for (int i = 0; i < 10; i++) {
            visited[i] = false;
        }
        dfs(0, "");
        System.out.println(answer.get(answer.size() - 1));
        System.out.println(answer.get(0));
    }

    public static void dfs(int cnt, String word) {
        if (word.length() == k) {
            answer.add(word);
            return;
        }

        if (cnt >= k) {
            return;
        }

        for (int i = 0; i < 10; i++) {
            if (!visited[i]) {
                if (word.length() < 1) {
                    visited[i] = true;
                    dfs(cnt + 1, word + i);
                    visited[i] = false;
                } else {
                    if (arr[cnt - 1].equals(">")) {
                        if (word.charAt(word.length() - 1) - '0' > i) {
                            visited[i] = true;
                            dfs(cnt + 1, word + i);
                            visited[i] = false;
                        }
                    } else {
                        if (word.charAt(word.length() - 1) - '0' < i) {
                            visited[i] = true;
                            dfs(cnt + 1, word + i);
                            visited[i] = false;
                        }
                    }
                }
            }
        }
    }
}

File: 백준/Silver/2529. 부등호/부등호.py
import sys
input = sys.stdin.readline


def dfs(cnt, word):
    if len(word) == k:
        answer.append(word)
        return

    if cnt >= k:
        return

    for i in range(10):
        if not visited[i]:
            if len(word) < 1:
                visited[i] = 1
                dfs(cnt + 1, word + str(i))
                visited[i] = 0
            else:
                if arr[cnt - 1] == ">":
                    if int(word[-1]) > i:
                        visited[i] = 1
                        dfs(cnt + 1, word + str(i))
                        visited[i] = 0
                else:
                    if int(word[-1]) < i:
                        visited[i] = 1
                        dfs(cnt + 1, word + str(i))
                        visited[i] = 0


k = int(input()) + 1
arr = list(input().split())
answer = []
visited = [0] * 10
dfs(0, "")
print(answer[-1])
print(answer[0])

